src/core/Akka/Event/LogMessage.cs

```diff
@@ -5,46 +5,133 @@
 // </copyright>
 //-----------------------------------------------------------------------
 
+using System;
+using System.Threading;
+
 namespace Akka.Event
 {
+    public interface ILogContents
+    {
+        public LogLevel LogLevel { get; }
+
+        public Exception Exception { get; }
+        
+        /// <summary>
+        /// The timestamp that this event occurred.
+        /// </summary>
+        public DateTime Timestamp { get; }
+
+        /// <summary>
+        /// The thread where this event occurred.
+        /// </summary>
+        public int ThreadId { get; }
+
+        /// <summary>
+        /// The source that generated this event.
+        /// </summary>
+        public LogSource LogSource { get; }
+
+        /// <summary>
+        /// Renders an underlying <see cref="LogEntry{TState}"/> in a non-generic fashion.
+        /// </summary>
+        /// <returns></returns>
+        string Format();
+    }
+    
     /// <summary>
-    /// Represents a log message which is composed of a format string and format args.
+    /// Raw data payload produced from any logging call.
     /// </summary>
-    public class LogMessage
+    /// <typeparam name="TState">The state for the specified log message.</typeparam>
+    public readonly struct LogEntry<TState> : ILogContents
```

Comment:
Any cast from struct to interface is a boxing, using struct here instead of class is a tradeoff?

Reply:
So this is something I'm debating - to make this whole system more performant, we have to either commit to a larger breaking change or backwards compatibility.

1. **More performance** - use generics for log state all the way down and discard the old `LogMessage` types. We can make the impact on the `ILoggingAdapter` side of things minimal but all custom back-end logging infrastructure __will be broken__. Changes needed to fix will be minor but will require new source code.
2. **Backwards compat** - this is what I opted for in this approach so far: preserving backwards compat to the extent that it's possible. That's why there's boxing and other issues.

I'm going to go down the more radical route and see what needs to be done here.

Reply:
Also, it matters  _where_ the boxing happens - we 100% don't care if it happens in the consuming logger actor. We care a lot if it happens in the foreground.